include/dlaf/solver/triangular/impl.h

```diff
@@ -519,6 +521,91 @@ void Triangular<backend, device, T>::call_LLN(comm::CommunicatorGrid grid, blas:
+
+  const comm::Index2D this_rank = grid.rank();
+
+  const matrix::Distribution& distr_a = mat_a.distribution();
+  const matrix::Distribution& distr_b = mat_b.distribution();
+
+  if (mat_b.size().isEmpty())
+    return;
+
+  constexpr std::size_t n_workspaces = 2;
+  common::RoundRobin<matrix::Panel<Coord::Col, T, D>> a_panels(n_workspaces, distr_a);
+  common::RoundRobin<matrix::Panel<Coord::Row, T, D>> b_panels(n_workspaces, distr_b);
+
+  for (SizeType k = mat_a.nrTiles().cols() - 1; k >= 0; --k) {
+    const GlobalTileIndex kk{k, k};
+
+    const LocalTileIndex kk_offset{distr_a.nextLocalTileFromGlobalTile<Coord::Row>(kk.row()),
+                                   distr_a.nextLocalTileFromGlobalTile<Coord::Col>(kk.col())};
+    const LocalTileIndex bt_offset(distr_b.nextLocalTileFromGlobalTile<Coord::Row>(kk.row() + 1), 0);
+
+    auto& a_panel = a_panels.nextResource();
+    auto& b_panel = b_panels.nextResource();
+
+    a_panel.setRangeStart(kk);
+
+    if (kk.row() == mat_a.nrTiles().rows() - 1) {
+      a_panel.setWidth(mat_a.tileSize(kk).cols());
+      b_panel.setHeight(mat_a.tileSize(kk).rows());
+    }
+
+    const auto rank_kk = distr_a.rankGlobalTile(kk);
+    if (this_rank.col() == rank_kk.col()) {
+      for (SizeType i_loc = kk_offset.row(); i_loc < distr_a.localNrTiles().rows(); ++i_loc) {
+        const LocalTileIndex ik{i_loc, kk_offset.col()};
+        a_panel.setTile(ik, mat_a.read(ik));
+      }
+    }
+    comm::broadcast(executor_mpi, rank_kk.col(), a_panel, mpi_row_task_chain);
+
+    // TODO priority
+    matrix::util::set0<backend>(thread_priority::normal, b_panel);
+
+    // TODO priority
+    for (const auto& ij : common::iterate_range2d(bt_offset, indexFromOrigin(distr_b.localNrTiles())))
+      gemmTrailingMatrixTile<backend>(thread_priority::normal, op, T(1) / alpha, a_panel.read_sender(ij),
+                                      mat_b.read_sender(ij), b_panel.readwrite_sender(ij));
+
+    for (const auto& idx : b_panel.iteratorLocal()) {
+      if (this_rank.row() == rank_kk.row())
+        comm::scheduleReduceRecvInPlace(executor_mpi, mpi_col_task_chain(), MPI_SUM, b_panel(idx));
+      else
+        comm::scheduleReduceSend(executor_mpi, rank_kk.row(), mpi_col_task_chain(), MPI_SUM,
+                                 b_panel.read(idx));
+    }
+
+    if (this_rank.row() == rank_kk.row()) {
+      for (SizeType j_loc = 0; j_loc < distr_b.localNrTiles().cols(); ++j_loc) {
+        const LocalTileIndex kj(kk_offset.row(), j_loc);
+        // TODO priority
+        const auto& priority = thread_priority::high;
+
+        dlaf::internal::whenAllLift(T(-1), b_panel.read_sender(kj), mat_b.readwrite_sender(kj)) |
+            tile::add(dlaf::internal::Policy<backend>(priority)) |
+            hpx::execution::experimental::detach();
+
+        trsmBPanelTile<backend>(priority, op, diag, alpha, a_panel.read_sender(kk_offset),
+                                mat_b.readwrite_sender(kj));
```

Comment:
@msimberg in this kind of context, would you suggest using a `let_value(mat_b.readwrite_sender)` and do both `tile::add` and the work of `trsmBPanelTile` internally?

IMHO is should be better from a continuation point of view, but knowing the (current?) limitations with `reference_wrapper` and template deduction, the code may become more verbose and uglier (e.g. we cannot use the pipe operator inside, and we have to create a `ex::then()` with a lambda that helps the "helps" the template deduction)

Reply:
I would probably not suggest that, simply because of the big reduction in clarity and separation of the operations. It may indeed lead to a slight improvement in runtimes but I doubt the effect would be huge. If I understand you correctly you're essentially suggesting to "fuse" the access to the tile in `mat_b` into a single operation, correct? It could be beneficial but I would assume there's quite a lot of other higher impact changes that would come before this.